File: docs/sliding_window/2444_hard/bounded_subarrays.cpp
#include <vector>
using namespace std;

long long countBoundedSubarrays(vector<int>& nums, int minK, int maxK) { // LeetCode Q.2444.
    long long boundedSubarraysCount = 0;

    int prevMinIdx = -1, prevMaxIdx = -1;

    int leftIdx = 0;
    for (int rightIdx = 0; rightIdx < nums.size(); rightIdx++) {
        int num = nums[rightIdx];

        if (num < minK || num > maxK) { // Reset.
            prevMinIdx = -1, prevMaxIdx = -1;
            leftIdx = rightIdx + 1;
            continue;
        }

        if (num == maxK)
            prevMaxIdx = rightIdx;
        if (num == minK)
            prevMinIdx = rightIdx;

        if (min(prevMinIdx, prevMaxIdx) >= 0)
            boundedSubarraysCount += min(prevMinIdx, prevMaxIdx) + 1 - leftIdx;
    }

    return boundedSubarraysCount;
}

File: docs/sliding_window/2444_hard/bounded_subarrays.py
def count_bounded_subarrays(nums: list[int], min_k: int, max_k: int):
    bounded_subarrays_count = 0

    prev_min_idx, prev_max_idx = -1, -1

    left_idx = 0
    for right_idx, num in enumerate(nums):
        if num < min_k or num > max_k:  # Reset.
            prev_min_idx, prev_max_idx = -1, -1
            left_idx = right_idx + 1
            continue

        if num == max_k:
            prev_max_idx = right_idx
        if num == min_k:
            prev_min_idx = right_idx

        if min(prev_min_idx, prev_max_idx) >= 0:
            bounded_subarrays_count += min(prev_min_idx, prev_max_idx) + 1 - left_idx

    return bounded_subarrays_count
